### PR TITLE
Improvements to lunac

### PR DESCRIPTION
**Pull request by mikusp**
This pull request adds a few options to the luna compiler:
* ability to specify a project as an argument
* file-watch mode to recompile the project on change to any .luna file inside
* exhaustive mode to compile all of the functions and methods in stdlib and compiled project

### CODE REVIEW

File: shell/src/Luna/Shell.hs [hunk 1]
```diff
@@ -139,33 +147,129 @@ stdlibPath = do
...
+            Left err -> putStrLn $ "Luna encountered runtime error: " ++ err
+            _        -> putStrLn "main finished"
+    Nothing -> putStrLn "Function main not found in module Main."
+
+shell :: Options -> IO ()
+shell opts = do
+    mainPath' <- maybe getCurrentDirectory return $ opts ^. projectDirectory
+    mainPath  <- Path.parseAbsDir mainPath'
+    let mainName = Project.getProjectName mainPath
+    stdPath   <- stdlibPath
```
Comment (piotrMocz) [inline shell/src/Luna/Shell.hs]: Purely a nitpick, but if `mainPath'` --> `mainPath`, then `stdPath'` --> `stdPath` :D Also, I think `Path.parseAbsDir =<< stdlibPath` would be even more readable (one man's opinion).

Comment (mikusp) [reply shell/src/Luna/Shell.hs]: I agree that it's annoying, but both are required in different places and combining respective variables into one would lengthen some of the latter lines, causing them to be split.

File: shell/src/Luna/Shell.hs [hunk 2]
```diff
@@ -139,33 +147,129 @@ stdlibPath = do
...
+
+instance Exception CompilationError where
+    displayException (CompilationError errors) =
+        convert $ Layout.concatLineBlock $ Layout.render $ formatErrors errors
+
+data Options = Options { _projectDirectory      :: Maybe FilePath
+                       , _exhaustiveCompilation :: Bool
+                       , _fileWatch             :: Bool
+                       }
+
```
Comment (wdanilo) [inline shell/src/Luna/Shell.hs]: Improper code style

File: shell/src/Luna/Shell.hs [hunk 3]
```diff
@@ -139,33 +147,129 @@ stdlibPath = do
...
 main :: IO ()
-main = do
-    mainPath'    <- getCurrentDirectory
-    mainPath     <- Path.parseAbsDir mainPath'
-    let mainName  = Project.getProjectName mainPath
-    stdPath      <- stdlibPath
-    (_, std)     <- Project.prepareStdlib  (Map.fromList [("Std", stdPath)])
+main = shell =<< OptParse.execParser opts
+    where
```
Comment (wdanilo) [inline shell/src/Luna/Shell.hs]: could we move this `where` to previous like? In all the code we keep one of two styles:

```
foo bar baz where
   ...
```

or

```
foo bar baz
    where ...
          ...
```

File: shell/src/Luna/Shell.hs (hunk 1 above)
Comment (wdanilo) [reply shell/src/Luna/Shell.hs]: Maybe we should give them better names? By looking at the code fast I dont know what they are for.

File: shell/src/Luna/Shell.hs [hunk 4]
```diff
@@ -139,33 +147,129 @@ stdlibPath = do
...
+    Nothing -> putStrLn "Function main not found in module Main."
+
+shell :: Options -> IO ()
+shell opts = do
+    mainPath' <- maybe getCurrentDirectory return $ opts ^. projectDirectory
+    mainPath  <- Path.parseAbsDir mainPath'
+    let mainName = Project.getProjectName mainPath
+    stdPath   <- stdlibPath
+    stdPath'  <- Path.parseAbsDir stdPath
+    (fin, std) <- Project.prepareStdlib  (Map.fromList [("Std", stdPath)])
```
Comment (wdanilo) [inline shell/src/Luna/Shell.hs]: `Map.fromList` -> `fromList` (exported in Prologue)

File: shell/src/Luna/Shell.hs [hunk 5]
```diff
@@ -139,33 +147,129 @@ stdlibPath = do
...
+                e <- getLine
+                unless (e `P.elem` (["exit", "quit"] :: [P.String])) $ do
+                    atomically $ writeTVar dirtyVar True
+                    watchInput
+
+        FSNotify.withManager $ \mgr -> do
+            let predicate ev =
+                    FilePath.takeExtension (FSNotify.eventPath ev) == ".luna" 
+            FSNotify.watchTree mgr mainPath' predicate
+                (\_ -> atomically $ writeTVar dirtyVar True)
```
Comment (wdanilo) [inline shell/src/Luna/Shell.hs]: why not use `const` here?

File: shell/src/Luna/Shell.hs [hunk 6]
```diff
@@ -139,33 +147,129 @@ stdlibPath = do
...
+            race_ watchInput $ forever $ do
+                atomically $ do
+                    dirty <- readTVar dirtyVar
+                    check dirty
+
+                (fin >> loop) `Exc.catchAny` (putStrLn . displayException)
+
+                atomically $ writeTVar dirtyVar False
+
+                putStrLn "Build finished. Press Enter to force a rebuild"
```
Comment (wdanilo) [inline shell/src/Luna/Shell.hs]: <3

File: shell/src/Luna/Shell.hs [hunk 7]
```diff
@@ -6,6 +6,10 @@ module Luna.Shell where
 
 import           Luna.Prelude        hiding (String, seq, cons, Constructor)
 import qualified Luna.Prelude        as P
+import           Control.Concurrent  (threadDelay)
+import           Control.Concurrent.Async (race_)
+import           Control.Concurrent.STM
```
Comment (wdanilo) [inline shell/src/Luna/Shell.hs]: Could you please enumerate things youre using from this import or use qualified import instead? I know we used to code in such style, but lets try to keep imports either qualified or explicitly listed.

**Pull request merged**